### PR TITLE
KAFKA-17077: The node.id is inconsistent to broker.id when "broker.id.generation.enable=true".

### PR DESCRIPTION
**Pull request by frankvicky**
We directly change the `broker.id` in `KafkaConfig` when `broker.id.generation.enable` is set to `true`. However, this update is NOT synchronized with the `node.id` in `KafkaConfig`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaServer.scala [hunk 1]
```diff
@@ -259,6 +259,7 @@ class KafkaServer(
 
         /* generate brokerId */
         config.brokerId = getOrGenerateBrokerId(initialMetaPropsEnsemble)
+        config.nodeId = config.brokerId
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaServer.scala]: Could you please add comments to explain why zk broker needs to sync node.id here

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for this patch

File: core/src/main/scala/kafka/server/KafkaServer.scala (hunk 1 above)
Comment (frankvicky) [reply core/src/main/scala/kafka/server/KafkaServer.scala]: Sure.

**Comment by frankvicky** [comment]
Hi @chia7712 
I have add some comments to explain why we need to make `nodeId` synchronizes` with `brokerId`, PTAL 😄

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 2]
```diff
@@ -308,7 +308,9 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
   val brokerIdGenerationEnable: Boolean = getBoolean(ServerConfigs.BROKER_ID_GENERATION_ENABLE_CONFIG)
   val maxReservedBrokerId: Int = getInt(ServerConfigs.RESERVED_BROKER_MAX_ID_CONFIG)
   var brokerId: Int = getInt(ServerConfigs.BROKER_ID_CONFIG)
-  val nodeId: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
+  // Making `nodeId` mutable is a workaround to fix the issue of de-synchronization between the generated `brokerId` and `nodeId`.
+  // It should be back to immutable after removing zk.
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Could you please file jira for it and then remove this comment?

**Comment by frankvicky** [comment]
Hi @chia7712 
I have removed the comment and file a jira [ticket](https://issues.apache.org/jira/browse/KAFKA-17128), PTAL

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -308,7 +308,7 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
   val brokerIdGenerationEnable: Boolean = getBoolean(ServerConfigs.BROKER_ID_GENERATION_ENABLE_CONFIG)
   val maxReservedBrokerId: Int = getInt(ServerConfigs.RESERVED_BROKER_MAX_ID_CONFIG)
   var brokerId: Int = getInt(ServerConfigs.BROKER_ID_CONFIG)
-  val nodeId: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
+  var nodeId: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Could you please add access limit to node id and broker id? for example:
```scala
  private[server] var _nodeIf: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
  def nodeId: Int = _nodeIf
```

That can't protect us 100% from bugs, but 50% at least I think :smile:

**Comment by frankvicky** [comment]
Hi @chia7712 
I have make a little changes based on your feedback, PTAL 😄

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -308,7 +308,8 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
   val brokerIdGenerationEnable: Boolean = getBoolean(ServerConfigs.BROKER_ID_GENERATION_ENABLE_CONFIG)
   val maxReservedBrokerId: Int = getInt(ServerConfigs.RESERVED_BROKER_MAX_ID_CONFIG)
   var brokerId: Int = getInt(ServerConfigs.BROKER_ID_CONFIG)
-  val nodeId: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
+  private[server] var _nodeIf: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Could you apply this change to `brokerId` too?

**Review by chia7712 — COMMENTED** [review]
@frankvicky thanks for updated PR

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (frankvicky) [reply core/src/main/scala/kafka/server/KafkaConfig.scala]: Ah, I miss this, I will add it.

**Comment by frankvicky** [comment]
Hi @chia7712 
I have changed the scope of `brokerId`, PTAL

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 5]
```diff
@@ -307,8 +307,10 @@ class KafkaConfig private(doLog: Boolean, val props: util.Map[_, _])
   /** ********* General Configuration ***********/
   val brokerIdGenerationEnable: Boolean = getBoolean(ServerConfigs.BROKER_ID_GENERATION_ENABLE_CONFIG)
   val maxReservedBrokerId: Int = getInt(ServerConfigs.RESERVED_BROKER_MAX_ID_CONFIG)
-  var brokerId: Int = getInt(ServerConfigs.BROKER_ID_CONFIG)
-  val nodeId: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
+  private[server] var _brokerIf: Int = getInt(ServerConfigs.BROKER_ID_CONFIG)
+  def brokerId: Int = _brokerIf
+  private[server] var _nodeIf: Int = getInt(KRaftConfigs.NODE_ID_CONFIG)
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: `_nodeId` and `_brokerId`

**Comment by frankvicky** [comment]
Hi @chia7712 
I have correct the field name, PTAL

**Comment by chia7712** [comment]
> java.lang.NullPointerException
	at kafka.server.KafkaServerTest.testGeneratedBrokerIdSyncWithNodeId(KafkaServerTest.scala:205)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at java.util.ArrayList.forEach(ArrayList.java:1259)

Please fix the failed test

**Comment by frankvicky** [comment]
Hi @chia7712 
I have refactored the test case with `TestUtils`, PTAL 🐧

File: core/src/test/scala/unit/kafka/server/KafkaServerTest.scala [hunk 6]
```diff
@@ -178,6 +176,21 @@ class KafkaServerTest extends QuorumTestHarness {
     server.shutdown()
   }
 
+  @Test
+  def testGeneratedBrokerIdSyncWithNodeId(): Unit = {
```
Comment (chia7712) [inline core/src/test/scala/unit/kafka/server/KafkaServerTest.scala:180]: We can re-order the creation to remove the null check. for example:
```scala
    val props = TestUtils.createBrokerConfig(-1, zkConnect)
    props.put("reserved.broker.max.id", "2000")
    val kafkaServer = TestUtils.createServer(KafkaConfig.fromProps(props))
    try {
      kafkaServer.startup()
      assertEquals(2001, kafkaServer.config.brokerId)
      assertEquals(2001, kafkaServer.config.nodeId)
    } finally kafkaServer.shutdown()
```

**Comment by frankvicky** [comment]
Hi @chia7712 
I have rearrange the code structure to avoid unnecessary null check, PTAL 🐧

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**